File: apps/apple/core/Sources/LorvexWorkflow/TaskRecurrence.swift
import Foundation
import GRDB
import LorvexDomain
import LorvexStore

/// Top-level `set_task_recurrence` workflow op. Canonicalizes the user's
/// recurrence rule into the stored JSON shape, normalizes it through
/// ``ValidationRecurrence/normalizeTaskRecurrence(_:)``, and delegates
/// to ``RecurrenceConfig/applyRecurrenceChange(_:taskId:recurrencePatch:duePatch:today:version:now:)``
/// for the LWW-gated UPDATE + dependent column reconciliation.
public enum TaskRecurrence {
  /// User-facing rule shape; `freq` is the only required field, all
  /// others extend the canonical RRULE form.
  public struct RuleInput: Sendable, Equatable {
    public var freq: String
    public var interval: UInt32?
    public var byday: [String]?
    public var bymonth: [Int64]?
    public var bymonthday: [Int64]
    public var bysetpos: [Int64]?
    public var wkst: String?
    public var until: String?
    public var count: UInt32?
    /// Lorvex `ANCHOR` extension: `"completion"` repeats INTERVAL units after
    /// the task is completed; `nil`/`"schedule"` is the default fixed cadence.
    public var anchor: String?

    public init(
      freq: String,
      interval: UInt32? = nil,
      byday: [String]? = nil,
      bymonth: [Int64]? = nil,
      bymonthday: [Int64] = [],
      bysetpos: [Int64]? = nil,
      wkst: String? = nil,
      until: String? = nil,
      count: UInt32? = nil,
      anchor: String? = nil
    ) {
      self.freq = freq
      self.interval = interval
      self.byday = byday
      self.bymonth = bymonth
      self.bymonthday = bymonthday
      self.bysetpos = bysetpos
      self.wkst = wkst
      self.until = until
      self.count = count
      self.anchor = anchor
    }
  }

  /// Input envelope for ``setTaskRecurrence(_:hlc:input:)``.
  public struct SetTaskRecurrenceInput: Sendable {
    public var taskId: TaskId
    public var rule: RuleInput

    public init(taskId: TaskId, rule: RuleInput) {
      self.taskId = taskId
      self.rule = rule
    }
  }

  /// Result of ``setTaskRecurrence(_:hlc:input:)``: the enriched
  /// before / after task JSON plus a human-readable summary string.
  public struct MutationResult: Sendable {
    public let taskId: String
    public let beforeTask: JSONValue
    public let afterTask: JSONValue
    public let summary: String
  }

  /// Apply the recurrence rule to the task. The caller owns the
  /// surrounding transaction (the underlying
  /// ``RecurrenceConfig/applyRecurrenceChange`` helper opens its
  /// own immediate transaction when handed a writer; this entry takes a
  /// mid-transaction `Database` and delegates to the in-tx variant).
  public static func setTaskRecurrence(
    _ db: Database,
    hlc: HlcSession,
    input: SetTaskRecurrenceInput
  ) throws -> MutationResult {
    let freqLabel = try canonicalFreq(input.rule.freq)
    let rawJSON = try ruleJSONString(input.rule)
    let normalized: String?
    switch ValidationRecurrence.normalizeTaskRecurrence(rawJSON) {
    case .success(let value): normalized = value
    case .failure(let err): throw StoreError.validation(err.description)
    }
    guard let recurrenceJSON = normalized else {
      throw StoreError.validation(
        "recurrence rule resulted in empty after normalization")
    }

    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    let title = TaskResponse.taskTitle(before)
    let now = SyncTimestampFormat.syncTimestampNow()
    let today = try WorkflowTimezone.todayYmdForConn(db)
    let version = hlc.nextVersionString()

    do {
      _ = try RecurrenceConfig.applyRecurrenceChangeInTx(
        db,
        taskId: input.taskId,
        recurrencePatch: .set(recurrenceJSON),
        dueDatePatch: .unset,
        today: today,
        version: version,
        now: now)
    } catch let error as RecurrenceConfig.ChangeError {
      throw mapChangeError(error)
    }

    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    let summary = recurrenceSummary(
      title: title, rule: input.rule, freqLabel: freqLabel)
    return MutationResult(
      taskId: input.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary: summary)
  }

  // MARK: - Helpers

  private static func canonicalFreq(_ raw: String) throws -> String {
    switch raw.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() {
    case "DAILY": return "DAILY"
    case "WEEKLY": return "WEEKLY"
    case "MONTHLY": return "MONTHLY"
    case "YEARLY": return "YEARLY"
    default:
      throw StoreError.validation(
        "freq must be one of daily, weekly, monthly, yearly")
    }
  }

  private static func ruleJSONString(_ rule: RuleInput) throws -> String {
    if let byday = rule.byday, byday.isEmpty {
      throw StoreError.validation(
        "BYDAY array must contain at least one weekday code (or be omitted)")
    }
    var obj: [String: JSONValue] = [:]
    obj["FREQ"] = .string(try canonicalFreq(rule.freq))
    if let interval = rule.interval {
      obj["INTERVAL"] = .int(Int64(interval))
    }
    if let byday = rule.byday, !byday.isEmpty {
      obj["BYDAY"] = .array(byday.map(JSONValue.string))
    }
    if let bymonth = rule.bymonth, !bymonth.isEmpty {
      obj["BYMONTH"] = .array(bymonth.map(JSONValue.int))
    }
    if !rule.bymonthday.isEmpty {
      obj["BYMONTHDAY"] = .array(rule.bymonthday.map(JSONValue.int))
    }
    if let bysetpos = rule.bysetpos, !bysetpos.isEmpty {
      obj["BYSETPOS"] = .array(bysetpos.map(JSONValue.int))
    }
    if let wkst = rule.wkst {
      obj["WKST"] = .string(wkst)
    }
    if let until = rule.until {
      obj["UNTIL"] = .string(until)
    }
    if let count = rule.count {
      obj["COUNT"] = .int(Int64(count))
    }
    if let anchor = rule.anchor, !anchor.isEmpty {
      obj["ANCHOR"] = .string(anchor)
    }
    return try canonicalizeJSON(.object(obj))
  }

  private static func recurrenceSummary(
    title: String, rule: RuleInput, freqLabel: String
  ) -> String {
    let intervalPart: String
    if let n = rule.interval, n > 1 {
      intervalPart = " every \(n)"
    } else {
      intervalPart = ""
    }
    let bydayPart: String
    if let byday = rule.byday, !byday.isEmpty {
      bydayPart = " on \(byday.joined(separator: ","))"
    } else {
      bydayPart = ""
    }
    let bymonthdayPart: String
    if !rule.bymonthday.isEmpty {
      bymonthdayPart = " on day \(rule.bymonthday.map(String.init).joined(separator: ","))"
    } else {
      bymonthdayPart = ""
    }
    let countPart: String
    if let c = rule.count {
      countPart = " for \(c) occurrences"
    } else {
      countPart = ""
    }
    let untilPart: String
    if let u = rule.until {
      untilPart = " until \(u)"
    } else {
      untilPart = ""
    }
    let anchorPart = rule.anchor == "completion" ? " after completion" : ""
    return
      "Set recurrence on '\(title)': "
      + "\(freqLabel)\(intervalPart)\(bydayPart)\(bymonthdayPart)\(countPart)\(untilPart)\(anchorPart)"
  }

  private static func mapChangeError(_ error: RecurrenceConfig.ChangeError)
    -> StoreError
  {
    switch error {
    case .clearDueDateOnRecurring:
      return .validation("recurring tasks must have a due_date")
    case .staleVersion(let taskId):
      return .staleVersion(entity: EntityName.task, id: taskId)
    }
  }
}

File: apps/tauri/lorvex-workflow/src/task_recurrence.rs
use lorvex_domain::hlc_session::HlcSession;
use lorvex_domain::naming::ENTITY_TASK;
use lorvex_domain::TaskId;
use lorvex_store::StoreError;
use rusqlite::Connection;
use serde_json::{json, Map, Value};

use crate::recurrence_config::{DueAtPatch, RecurrenceChangeError};
use crate::task_response::{load_enriched_task_json, task_title};

#[derive(Debug, Clone)]
pub struct TaskRecurrenceRuleInput {
    pub freq: String,
    pub interval: Option<u32>,
    pub byday: Option<Vec<String>>,
    pub bymonth: Option<Vec<i64>>,
    pub bymonthday: Option<Vec<i64>>,
    pub bysetpos: Option<Vec<i64>>,
    pub wkst: Option<String>,
    pub until: Option<String>,
    pub count: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct SetTaskRecurrenceInput {
    pub task_id: TaskId,
    pub rule: TaskRecurrenceRuleInput,
}

#[derive(Debug, Clone)]
pub struct TaskRecurrenceMutationResult {
    pub task_id: String,
    pub before_task: Value,
    pub after_task: Value,
    pub summary: String,
}

fn recurrence_change_error_to_store(error: RecurrenceChangeError) -> StoreError {
    match error {
        RecurrenceChangeError::ClearDueDateOnRecurring => {
            StoreError::Validation("recurring tasks must have a due_date".to_string())
        }
        RecurrenceChangeError::DueTimeWithoutDueDate => StoreError::Validation(
            "due_time without due_date is invalid: a clock time requires a calendar day"
                .to_string(),
        ),
        RecurrenceChangeError::Db(error) => StoreError::from(error),
        RecurrenceChangeError::TransactionWrap(message) => {
            StoreError::Invariant(format!("transaction wrapper failure: {message}"))
        }
        RecurrenceChangeError::StaleVersion { task_id } => StoreError::StaleVersion {
            entity: ENTITY_TASK,
            id: task_id,
        },
    }
}

fn canonical_freq(raw: &str) -> Result<&'static str, StoreError> {
    match raw.trim().to_ascii_uppercase().as_str() {
        "DAILY" => Ok("DAILY"),
        "WEEKLY" => Ok("WEEKLY"),
        "MONTHLY" => Ok("MONTHLY"),
        "YEARLY" => Ok("YEARLY"),
        _ => Err(StoreError::Validation(
            "freq must be one of daily, weekly, monthly, yearly".to_string(),
        )),
    }
}

fn to_rule_json_string(rule: &TaskRecurrenceRuleInput) -> Result<String, StoreError> {
    if rule.byday.as_ref().is_some_and(Vec::is_empty) {
        return Err(StoreError::Validation(
            "BYDAY array must contain at least one weekday code (or be omitted)".to_string(),
        ));
    }

    let mut object = Map::new();
    object.insert(
        "FREQ".to_string(),
        Value::String(canonical_freq(&rule.freq)?.to_string()),
    );
    if let Some(interval) = rule.interval {
        object.insert("INTERVAL".to_string(), json!(interval));
    }
    if let Some(byday) = rule.byday.as_ref().filter(|values| !values.is_empty()) {
        object.insert(
            "BYDAY".to_string(),
            Value::Array(byday.iter().cloned().map(Value::String).collect()),
        );
    }
    if let Some(bymonth) = rule.bymonth.as_ref().filter(|values| !values.is_empty()) {
        object.insert("BYMONTH".to_string(), json!(bymonth));
    }
    if let Some(bymonthday) = rule.bymonthday.as_ref().filter(|values| !values.is_empty()) {
        object.insert("BYMONTHDAY".to_string(), json!(bymonthday));
    }
    if let Some(bysetpos) = rule.bysetpos.as_ref().filter(|values| !values.is_empty()) {
        object.insert("BYSETPOS".to_string(), json!(bysetpos));
    }
    if let Some(wkst) = &rule.wkst {
        object.insert("WKST".to_string(), Value::String(wkst.clone()));
    }
    if let Some(until) = &rule.until {
        object.insert("UNTIL".to_string(), Value::String(until.clone()));
    }
    if let Some(count) = rule.count {
        object.insert("COUNT".to_string(), json!(count));
    }
    Ok(Value::Object(object).to_string())
}

fn recurrence_summary(title: &str, rule: &TaskRecurrenceRuleInput, freq_label: &str) -> String {
    let interval_part = if rule.interval.unwrap_or(1) > 1 {
        format!(" every {}", rule.interval.unwrap_or(1))
    } else {
        String::new()
    };
    let byday_part = rule
        .byday
        .as_ref()
        .filter(|values| !values.is_empty())
        .map(|values| format!(" on {}", values.join(",")))
        .unwrap_or_default();
    let bymonthday_part = rule
        .bymonthday
        .as_ref()
        .filter(|values| !values.is_empty())
        .map(|days| {
            let joined = days
                .iter()
                .map(i64::to_string)
                .collect::<Vec<_>>()
                .join(",");
            format!(" on day {joined}")
        })
        .unwrap_or_default();
    let count_part = rule
        .count
        .map(|count| format!(" for {count} occurrences"))
        .unwrap_or_default();
    let until_part = rule
        .until
        .as_ref()
        .map(|until| format!(" until {until}"))
        .unwrap_or_default();
    format!(
        "Set recurrence on '{title}': {freq_label}{interval_part}{byday_part}{bymonthday_part}{count_part}{until_part}"
    )
}

pub fn set_task_recurrence(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: SetTaskRecurrenceInput,
) -> Result<TaskRecurrenceMutationResult, StoreError> {
    let SetTaskRecurrenceInput { task_id, rule } = input;
    let freq_label = canonical_freq(&rule.freq)?;
    let raw_json = to_rule_json_string(&rule)?;
    let recurrence_json = lorvex_domain::validation::normalize_task_recurrence(&raw_json)?
        .ok_or_else(|| {
            StoreError::Validation(
                "recurrence rule resulted in empty after normalization".to_string(),
            )
        })?;

    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let now = lorvex_domain::sync_timestamp_now();
    let today = crate::timezone::today_ymd_for_conn(conn)?;
    let version = hlc.next_version_string();
    crate::recurrence_config::apply_recurrence_change(
        conn,
        &task_id,
        lorvex_domain::Patch::Set(recurrence_json),
        DueAtPatch::not_present(),
        &today,
        &version,
        &now,
    )
    .map_err(recurrence_change_error_to_store)?;

    let after = load_enriched_task_json(conn, &task_id)?;
    let summary = recurrence_summary(&title, &rule, freq_label);
    Ok(TaskRecurrenceMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary,
    })
}
